### PR TITLE
adding add_registry_creds for anchore

### PR DESCRIPTION
**Pull request by jpke**
# PR Details

Adds method `add_registry_creds` in Anchore library. This method will add to Anchore private docker registry credentials, allowing Anchore to pull and scan images from this repository.

This PR also changes the `git push` to `git "push"` in the `deploy_to` step of the Kubernetes library. This is done because 'push' of `git push` no longer appears to be found.

## Description

Adding an optional `add_registry_creds` step in the Anchore library, with additional optional configuration items: 

```
        docker_registry_credential_id = String
        docker_registry_name = String
        k8s_credential = String
        k8s_context = String
```

These changes are opt-in, being fully reverse compatible with Anchore library usage prior to the commit.

This PR also includes updates to the Anchore documentation page.

## How Has This Been Tested

This new method has been tested in an sdp pipeline, on both fresh environments, and on repeated pipeline runs.

## Types of Changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I am submitting this pull request to the appropriate branch
- [x] I have labeled this pull request appropriately
- [x] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            registry: "$registry",
+            registry_name: "$registry",
+            registry_type: 'docker_v2',
+            registry_user: "$registry_user",
+            registry_verify: true,
+            registry_pass: "$registry_password"
+        ])
+}
+
+void call(app_env){
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:21]: if the `app_env` is optional, perhaps this should be have a default parameter so that `add_registry_creds()` could be called without application environment. 

```suggestion
void call(app_env = null){
```

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    stage("Ensure Anchore has Docker Creds"){
+
+        /*
+        Docker registry creds for pulling helm image (has kubectl)
+        */
+        def docker_registry_credential_id = app_env && app_env.docker_registry_credential_id ?:
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:27]: with the null method parameter, you can use the [Safe Access Operator](https://blog.mrhaki.com/2009/08/groovy-goodness-safe-navigation-to.html) to avoid a null pointer exception. 

```suggestion
        def docker_registry_credential_id = app_env?.docker_registry_credential_id ?: 
```

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    stage("Ensure Anchore has Docker Creds"){
+
+        /*
+        Docker registry creds for pulling helm image (has kubectl)
+        */
+        def docker_registry_credential_id = app_env && app_env.docker_registry_credential_id ?: 
+                                                config.docker_registry_credential_id ?: 
+                                                "docker_registry"
+                                                
+        def docker_registry_name = app_env && app_env.docker_registry_name ?:
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:31]:
```suggestion
        def docker_registry_name = app_env?.docker_registry_name ?: 
```

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                                                "docker_registry"
+                                                
+        def docker_registry_name = app_env && app_env.docker_registry_name ?: 
+                                        config.docker_registry_name ?: 
+                                        ""
+
+        /*
+        k8s credential with kubeconfig 
+        */
+        def k8s_credential = app_env && app_env.k8s_credential ?:
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:38]:
```suggestion
        def k8s_credential = app_env?.k8s_credential ?:
```

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                                config.k8s_credential  ?:
+                                {error "Kubernetes Credential Not Defined"}()
+        /*
+        k8s context
+        */
+        def k8s_context = app_env && app_env.k8s_context ?:
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:44]:
```suggestion
        def k8s_context = app_env?.k8s_context ?:
```

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                            config.k8s_context ?:
+                            {error "Kubernetes Context Not Defined"}()
+
+        /*
+        Anchore engine url
+        */
+        env.ANCHORE_ENGINE_URL = app_env && app_env.anchore_engine_url ?:
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:51]:
```suggestion
        env.ANCHORE_ENGINE_URL = app_env?.anchore_engine_url ?: 
```

File: libraries/anchore/steps/add_registry_creds.groovy [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                            {error "Kubernetes Context Not Defined"}()
+
+        /*
+        Anchore engine url
+        */
+        env.ANCHORE_ENGINE_URL = app_env && app_env.anchore_engine_url ?: 
+                                    config.anchore_engine_url ?: 
+                                    {error "Anchore Engine Url Not Defined"}()
+
+        node{
```
Comment (steven-terrana) [inline libraries/anchore/steps/add_registry_creds.groovy:55]: i don't think you need this node block because `inside_sdp_image` is a wrapper around `docker.image(--).inside{}` which creates a node block for you. 

The source code that shows that is [here](https://github.com/jenkinsci/docker-workflow-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/docker/workflow/Docker.groovy#L113-L130)

**Review by steven-terrana — CHANGES_REQUESTED** [review]
Happy to discuss more if needed!

only bigger picture feedback was wondering if this should be a `@BeforeStep` annotated step so that you don't have to call it explicitly in the template and it just runs prior to `container_image_scan()`

**Comment by steven-terrana** [comment]
i'm fine with deferring it until it's actually a pressing concern. 

good catch on the existing implementations. 

you'd handle that with lifecycle hooks conditional execution feature:

if you put all the registry creds in a `creds` config block you could do:

```
@BeforeStep({ config.containsKey("creds") })
```

but given it takes an `ApplicationEnvironment` you'd also have to check if the step received a param and check that too lol

this would be a new feature for JTE to access step params from a hook but it'd look like:

```
@BeforeStep({ config.containsKey("creds") || hookContext.stepArgs.first()?.anchore?.creds })
```

assuming the config on application environments was:

```
application_environments{
  dev{
    anchore{
      creds{ /* where you'd define these params */ }
    }
  }
}
```

cc @cokieffebah on the feature idea for hooks.

IN SUMMARY lol - it's probably not worth the complexity 👍

**Comment by jpke** [comment]
@steven-terrana anything else required on my end to prep this for merging?

**Review by steven-terrana — APPROVED** [review]
